packages/flutter_tools/lib/src/base/file_system.dart

```diff
@@ -136,11 +142,89 @@ class FileSystemUtils {
   /// Return the absolute path of the user's home directory
   String get homeDirPath {
     String path = _platform.isWindows
-        ? _platform.environment['USERPROFILE']
-        : _platform.environment['HOME'];
+      ? _platform.environment['USERPROFILE']
+      : _platform.environment['HOME'];
     if (path != null) {
       path = _fileSystem.path.absolute(path);
     }
     return path;
   }
 }
+
+/// This class extends [local_fs.LocalFileSystem] in order to clean up
+/// directories and files that the tool creates under the system temporary
+/// directory when the tool exits either normally or when killed by a signal.
+class LocalFileSystem extends local_fs.LocalFileSystem {
+  LocalFileSystem._(Signals signals, List<ProcessSignal> fatalSignals) :
+    _signals = signals, _fatalSignals = fatalSignals;
+
+  @visibleForTesting
+  LocalFileSystem.test({
+    @required Signals signals,
+    List<ProcessSignal> fatalSignals = Signals.defaultExitSignals,
+  }) : this._(signals, fatalSignals);
+
+  // Unless we're in a test of this class's signal hanlding features, we must
+  // have only one instance created with the singleton LocalSignals instance
+  // and the catchable signals it considers to be fatal.
+  static LocalFileSystem _instance;
+  static LocalFileSystem get instance => _instance ??= LocalFileSystem._(
+    LocalSignals.instance,
+    Signals.defaultExitSignals,
+  );
+
+  Directory _systemTemp;
+  final Map<ProcessSignal, Object> _signalTokens = <ProcessSignal, Object>{};
+
+  @visibleForTesting
+  static Future<void> dispose() => LocalFileSystem.instance?._dispose();
+
+  Future<void> _dispose() async {
+    _tryToDeleteTemp();
+    for (final MapEntry<ProcessSignal, Object> signalToken in _signalTokens.entries) {
+      await _signals.removeHandler(signalToken.key, signalToken.value);
+    }
+    _signalTokens.clear();
+  }
+
+  final Signals _signals;
+  final List<ProcessSignal> _fatalSignals;
+
+  void _tryToDeleteTemp() {
```

Comment:
It looks like we're blowing away the entire temp here, is that safe to do? Are there issues with tracking the individual temp entries we create?

Reply:
It's not the entire system temp. Maybe it's a bad variable name. Below, the `systemTempDirectory` getter returns a fresh entry under `/tmp`, like `/tmp/flutter_tools.abcxyz`, then the rest of the tool creates `/tmp` entries under that, like `/tmp/flutter_tools.abcxyz/flutter_build_stuff.123456`. Right before exiting because of a signal or otherwise, this code deletes `/tmp/flutter_tools.abcxyz`.

Reply:
Ahh okay, that makes more sense.

Reply:
Promoted my response to a code comment.

Reply:
One potential issue on windows is that we may fail to delete if there is a process holding onto just one of those files.

Reply:
... I think

Reply:
Yes, that can happen. That's the reason for the `try {} catch {}`. On the flip side, on other platforms, we may see integration test failures if the tool is pulling the rug out from underneath another process.